mathics/builtin/colors/color_directives.py

```diff
@@ -401,15 +404,15 @@ def apply(self, c1, c2, evaluation, options):
 
             def compute(a, b):
                 return Expression(
-                    "Apply",
+                    SymbolApply,
                     distance_function,
                     Expression(
-                        "List",
+                        SymbolList,
```

Comment:
@mmatera As I go over the conversion for the Expression interface I am seeing so many Literal constants so many places that, I suspect having that part in alone will be a big win. 

And I suspect we can get that in faster than going over all of the Expressions and sorting the whites from the colored (to use my favorite laundry analogy). 

I also think that neither precludes the other and both  Expression interface change and handling Lists will be an improvement on just one. 

The main thing I worry about is making sure Lists goes in knowing that underneath there are those 3 properties: fully evaluated, ordered, and flat.

Reply:
> @mmatera As I go over the conversion for the Expression interface I am seeing so many Literal constants so many places that, I suspect having that part in alone will be a big win.
> 


Do you mean, to use always the plane constructor (the one without the type conversion)? I agree with that.  But mainly, because we can avoid the constructor doing a lot of extra work in cases where the conversions are not needed, what happens in the rewrite_and_apply method, and (probably) the ``do_apply_rule``.


> And I suspect we can get that in faster than going over all of the Expressions and sorting the whites from the colored (to use my favorite laundry analogy).
> 

Yep, this would clarify more what is the landscape we have in front.

> I also think that neither precludes the other and both Expression interface change and handling Lists will be an improvement on just one.
> 
> The main thing I worry about is making sure Lists goes in knowing that underneath there are those 3 properties: fully evaluated, ordered, and flat.

I am not sure to understand this last point: I `List` where a different class, you shouldn´t worry about sorting:  `List` never have the attribute `Orderless`.  They also do not have the attribute `Flat`, but in that case, maybe you gain something if you know that a list is already flat when it is an element of another expression.

Reply:
> > @mmatera As I go over the conversion for the Expression interface I am seeing so many Literal constants so many places that, I suspect having that part in alone will be a big win.
> 
> Do you mean, to use always the plane constructor (the one without the type conversion)? I agree with that. But mainly, because we can avoid the constructor doing a lot of extra work in cases where the conversions are not needed, what happens in the rewrite_and_apply method, and (probably) the `do_apply_rule`.

There is a part where I agee, and another part where you may be mistaken. I will elaborate and explain below.

> 
> > And I suspect we can get that in faster than going over all of the Expressions and sorting the whites from the colored (to use my favorite laundry analogy).
> 
> Yep, this would clarify more what is the landscape we have in front.
> 
> > I also think that neither precludes the other and both Expression interface change and handling Lists will be an improvement on just one.
> > The main thing I worry about is making sure Lists goes in knowing that underneath there are those 3 properties: fully evaluated, ordered, and flat.
> 
> I am not sure to understand this last point: I `List` where a different class, you shouldn´t worry about sorting: `List` never have the attribute `Orderless`. They also do not have the attribute `Flat`, but in that case, maybe you gain something if you know that a list is already flat when it is an element of another expression.


Some operations like "Plus" are "Flat", "Orderless", and "Listable". While the properties of the elements of a List are not of use to the List function, it _is_ of use to the a "Plus" function, which it gets from querying the List elements structure.

I will make a stab at a ListExpression class over the weekend so as to make some of these ideas more concrete, and to insure that I convey the characteristics that I think are useful and important.

However in this stab, there will be some deliberate short-cuts and places for correction and improvement. Some of this is because I would like to have something working that we can iterate on, and already I see that as happens so often, since the code base is large and uses List everywhere it will probably be a chore to convert everything initially.

So I imagine some parts will be done and other parts will need to be done. But it will be clear how to do the remaining part. Similar to the situation with Symbol where we have many places where we are still using Symbol("X") instead of SymbolX.

Before doing this though, there will be some work done to adjust the expression-interface-refactor branch so that it can work with our currently unsorted (sloppy) use of Expression.  Right now that branch can't be in master because it rejects stuff that isn't sorted. And it will take a long time for Expressions to get sorted. 

The main change will be to get the interfaces adjusted and to start useing the ElementsProperties mutable record.

While the expression-interface-refactor branch can't be used in general, it is however very useful for getting a sense of the performance improvement we will see, because for small limited kinds of things, small benchmark test in particular, we this code does work. 

The other places that will need correction is that I intend to start off subclassing Expression which is too large. What I imagine will be needed is a BaseExpression that is stripped down that can serve as the the Base class for both (compound) Expression and ListExpression. In the future Associations and possibly Boxing kinds of things might also fit here. You probably know about that more than I.